metrics/serve.go

```diff
@@ -82,6 +82,135 @@ func cycleValues(labelKeys, labelValues []string, seriesCount, seriesCycle int)
 	}
 }
 
+func handleValueTicks(labelKeys, labelValues *[]string, currentSeriesCount, seriesCycle *int, updateNotify chan struct{}, valueTick *time.Ticker) {
+	for tick := range valueTick.C {
+		metricsMux.Lock()
+		fmt.Printf("%v: refreshing metric values\n", tick)
+		cycleValues(*labelKeys, *labelValues, *currentSeriesCount, *seriesCycle)
+		metricsMux.Unlock()
+
+		select {
+		case updateNotify <- struct{}{}:
+		default:
+		}
+	}
+
+}
+
+func handleSeriesTicks(labelKeys, labelValues *[]string, currentSeriesCount, seriesCycle *int, updateNotify chan struct{}, seriesTick *time.Ticker) {
+	for tick := range seriesTick.C {
+		metricsMux.Lock()
+		fmt.Printf("%v: refreshing series cycle\n", tick)
+		deleteValues(*labelKeys, *labelValues, *currentSeriesCount, *seriesCycle)
+		(*seriesCycle)++
+		cycleValues(*labelKeys, *labelValues, *currentSeriesCount, *seriesCycle)
+		metricsMux.Unlock()
+
+		select {
+		case updateNotify <- struct{}{}:
+		default:
+		}
+	}
+
+}
+
+func handleMetricTicks(metricCount, metricLength, metricCycle *int, labelKeys *[]string, updateNotify chan struct{}, metricTick *time.Ticker) {
+	for tick := range metricTick.C {
+		metricsMux.Lock()
+		fmt.Printf("%v: refreshing metric cycle\n", tick)
+		(*metricCycle)++
+		unregisterMetrics()
+		registerMetrics(*metricCount, *metricLength, *metricCycle, *labelKeys)
+		metricsMux.Unlock()
+		select {
+		case updateNotify <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func changeSeriesGradual(seriesChangeRate, maxSeriesCount, minSeriesCount, currentSeriesCount *int, seriesIncrease *bool) {
+	if *seriesIncrease {
+		*currentSeriesCount += *seriesChangeRate
+		if *currentSeriesCount >= *maxSeriesCount {
+			*currentSeriesCount = *maxSeriesCount
+			*seriesIncrease = false
+		}
+	} else {
+		*currentSeriesCount -= *seriesChangeRate
+		if *currentSeriesCount <= *minSeriesCount {
+			*currentSeriesCount = *minSeriesCount
+			*seriesIncrease = true
+		}
+	}
+}
+
+func changeSeriesDoubleHalve(currentSeriesCount *int, seriesIncrease *bool) {
+	if *seriesIncrease {
+		*currentSeriesCount *= 2
+	} else {
+		*currentSeriesCount /= 2
+		if *currentSeriesCount < 1 {
+			*currentSeriesCount = 1
+		}
+	}
+	*seriesIncrease = !*seriesIncrease
+}
+
+func handleDoubleHalveMode(metricCount, metricLength, metricCycle, seriesCycle int, labelKeys, labelValues []string, seriesCount int, changeSeriesTick *time.Ticker, updateNotify chan struct{}) {
+	currentSeriesCount := seriesCount
+	seriesIncrease := true
+
+	registerMetrics(metricCount, metricLength, metricCycle, labelKeys)
+	cycleValues(labelKeys, labelValues, currentSeriesCount, seriesCycle)
+
+	go func() {
```

Comment:
just for consistency, can we have this function and `handleGradualChangeMode` just do their loop, and move the `go func() { ... }` wrapper to where they are called in `RunMetrics`

Reply:
moved👍